### DISCUSSION TITLE
What is the best C# "codemod" engine out there?

### DISCUSSION THREAD

**Discussion by alexbit-codemod** [JosefPihrt/Roslynator#1026] [Q&A]
Hey everyone! 

Is there any tool (maybe Roslynator is the closest to what I'm looking for) to **help C# devs easily (maybe in a declarative way) create a script to detect certain patterns and transform it to something else (codemod)** so they can do their mass refactoring, migrations, and upgrades easier?

in JS/TS world we have: https://github.com/facebook/jscodeshift
and popular frameworks (e.g. Material UI) create codemods for their consumers to run and almost automatically upgrade their dependencies (see these for [MUI v4 to v5 upgrade](https://github.com/mui/material-ui/tree/master/packages/mui-codemod/src/v5.0.0).)

there are many other codemod engines out there.
LibCST for Python
openrewrite/rewrite for java
Uber's Piranha (for many languages) 
JScodeshit, or Putout for JS/TS

**Comment by josefpihrt** [comment]
Hi,

Roslynator command line tool can be used for this purpose (namely `roslynator fix` command) (https://github.com/JosefPihrt/Roslynator/blob/main/docs/cli/fix-command.md).

Also, `dotnet format` tool can be used (https://learn.microsoft.com/en-us/dotnet/core/tools/dotnet-format).

**Reply by alexbit-codemod** [reply]
thank you. do you know of any sample ruleset that helps with the upgrade of a popular c# framework from one version to another? i just want to find some existing code transformations rules... how to define a code patterns and how to replace it with something else.

**Reply by kaykaul** [reply]
@alexbit-codemod were you able to find a solution?

**Reply by alexbit-codemod** [reply]
no but ive learned that you can easily add support for any languages to Uber's Piranha. in what capacity are you interested in C# codemod engine? if you wanna contributes to Uber's codebase i have connections and guidance to add support for C#. its pretty straightforward. you can ping me here: https://go.codemod.com/slack
